lib/presto-client/index.js

```diff
@@ -220,167 +229,212 @@ Client.prototype.statementResource = function(opts) {
-                                 "totalSplits": 420,
-                                 "queuedSplits": 0,
-                                 "runningSplits": 0,
-                                 "completedSplits": 420
-                             }
-                         ],
-                         // same as substage
-                     },
-                     // same as substage
-                     "state": "RUNNING",
-                 },
-                 "data": [ [ 1266352 ] ],
-                 "columns": [ { "type": "bigint", "name": "cnt" } ],
-                 "nextUri": "http://localhost:8080/v1/statement/20140120_032523_00000_32v8g/2",
-                 "partialCancelUri": "http://10.0.0.0:8080/v1/stage/20140120_032523_00000_32v8g.0",
-                 "infoUri": "http://localhost:8080/v1/query/20140120_032523_00000_32v8g",
-                 "id": "20140120_032523_00000_32v8g"
-             }
-            */
-            /*
-             * 2nd time
-             {
-                 "stats": {
-                     // ....
-                     "state": "FINISHED",
-                 },
-                 "columns": [ { "type": "bigint", "name": "cnt" } ],
-                 "infoUri": "http://localhost:8080/v1/query/20140120_032523_00000_32v8g",
-                 "id": "20140120_032523_00000_32v8g"
-             }
-            */
-            if (cancel_checker && cancel_checker()) {
-                client.request({ method: 'DELETE', path: next_uri }, function(error, code, data){
-                    if (error || code !== 204) {
-                        error_callback({message: "query fetch canceled, but Presto query cancel may fail", error: error, code: code});
-                    } else {
-                        error_callback({message: "query fetch canceled by operation"});
-                    }
-                });
+        current_req = client.request(next_uri, function(error, code, response){
+            // same as above, we have already timed out and shown an error, so abort out
+            if (timed_out) {
+                return;
+            } else if ([502, 503, 504].includes(code)) {
+                setTimeout(function(){
+                    fetch_next(next_uri);
+                }, Math.floor(Math.random() * 51) + 50);
+                if (retry_callback) {
+                    retry_callback();
+                }
                 return;
             }
-            client.request(next_uri, function(error, code, response){
-                if (error || response.error) {
+
+            if (error || (response && response.error)) {
+                if (first_request) {
+                    var message = "execution error" + (response && response.length > 0 ? ":" + response : "");
+                    if (response && response.error && response.error.message)
+                        message = response.error.message;
+                    error_callback({message, error: (error || response.error), code});
```

Comment:
The error callback argument is modified to an incompatible form(?) only when it's the first request. What's the reason?

Reply:
This is combining the original logic of:
https://github.com/tagomoris/presto-client-node/blob/84f76d981482c5dd710a147ebfe89efa1731d85f/lib/presto-client/index.js#L230-L238

and

https://github.com/tagomoris/presto-client-node/blob/84f76d981482c5dd710a147ebfe89efa1731d85f/lib/presto-client/index.js#L333-L336

into a single function call, where the former was only done on the first POST request, while the latter was done for all subsequent GET requests.